olm/operator-registry.md

```diff
@@ -0,0 +1,291 @@
+
+Add a new set of operator registry commands to utilize those new APIs:
+
+`operator-registry create`
+    - inputs: none
+    - outputs: empty operator registry database
+    
+`operator-registry add`
+    - inputs: $operatorBundleImagePath, $dbFile
+    - outputs: updated database file
+    ex: `operator-registry add quay.io/community-operators/foo:sha256@abcd123 example.db`
+    
+`operator-registry add --batch`
+    - inputs: $operatorBundleImagePath, $dbFile
+    - outputs: updated database file
+    ex: `operator-registry add "quay.io/community-operators/foo:sha256@abcd123,quay.io/community-operators/bar:sha256@defg456" example.db`
+    
+`operator-registry delete`
+    - inputs: $operatorName, $dbFile
+    - outputs: updated database file without $operatorName included
+    ex: `operator-registry delete bar example.db`
+
+`operator-registry delete-latest`
+    - inputs: $operatorName, $dbFile
+    - outputs: updated database file without latest version of $operatorName included
+    ex: `operator-registry delete-latest foo example.db`
+    
+*Reference non latest versions of bundles by image digests*
+
+Currently the operator-registry database has a table to the effect of:
+
+OperatorBundle
+    -name: foo
+    -bundle: "{$jsonblob}"
+    
+We will add a field to this table that will reference the bundle image path and, when the operator is not the latest version of the default channel, will not include the bundle blob itself:
+
+OperatorBundle
+    -name: foo
+    -bundle: NULL
+    -bundlePath: "quay.io/community-operators/foo:sha256@abcd123"
+    
+Additionally, we will update the grpc API that OLM currently uses to get the bundle that if it attempts to pull the non latest version will return null. In that case, it will make a second query to get the bundlePath. Then we will create a pod with that image that will serve the bundle data by writing the data to a configmap that can be read by OLM.
+
+In order to serve that manifest image, we cannot pull the image directly. Instead we will:
+
+1. Create a pod spec that uses an init container to inject a binary into the manifest image.
+2. Run the manifest image with that binary that writes to a configmap in a well known location
+3. Read the configmap data
+4. On return kill the pod
+5. Delete the configmap
+
+pod spec:
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: example
+spec:
+  containers:
+  - name: operator-manifest
+    image: quay.io/community-operators/foo@sha256:abc123
+    command: ['operator-registry', 'serve' '-manifestdir "./manifests/"']
+  initContainers:
+  - name: init-manifest
+    image: quay.io/operator-framework/init-operator-manifest:latest
+    command: ['cp', 'operator-registry', '.']
+```
+
+We will also need to update operator-registry to include a new `serve` command that knows how to parse the manifest/ folder in the operator bundle image. This command will write to a configmap that OLM will use to get the bundle data.
+
+*DB Versioning*
```

Comment:
We should call out our versioning strategy for index images and CatalogSources in general.

Proposal: 

- DB format is versioned with schemas
- Tooling is versioned and supports a range of db schemas, indicated in `operator-registry/README.md`. 
- GRPC api follows [convention](https://developers.google.com/protocol-buffers/docs/proto#updating) and doesn't make backwards incompatible changes
- Any large changes we wish to make (i.e. want to deprecate the current grpc api because we've changed how dependency resolution works) will be indicated via new CatalogSource types (current types are `grpc`, and `configmap`/`internal`).